libyara/modules/macho.c

```diff
@@ -0,0 +1,1180 @@
+        break;                                                                 \
+      }                                                                        \
+      case LC_MAIN:                                                            \
+      {                                                                        \
+        macho_handle_main_##bo(command, object, context);                      \
+        break;                                                                 \
+      }                                                                        \
+    }                                                                          \
+    command += yr_##bo##32toh(command_struct->cmdsize);                        \
+  }                                                                            \
+                                                                               \
+  set_integer(num_sg, object, "number_of_segments");                           \
+}                                                                              \
+
+MACHO_PARSE_FILE(32,le)
+MACHO_PARSE_FILE(64,le)
+MACHO_PARSE_FILE(32,be)
+MACHO_PARSE_FILE(64,be)
+
+
+// Parse Mach-O file.
+
+void macho_parse_file(
+    void* data,
+    YR_OBJECT* object,
+    YR_SCAN_CONTEXT* context)
+{
+  uint8_t* magic = (uint8_t*)data;
+  if (macho_is_32(magic))
+  {
+    macho_is_big(magic) ? macho_parse_file_32_be(data, object, context)
+                        : macho_parse_file_32_le(data, object, context);
+  }
+  else
+  {
+    macho_is_big(magic) ? macho_parse_file_64_be(data, object, context)
+                        : macho_parse_file_64_le(data, object, context);
+  }
+}
+
+
+// Parse Mach-O fat file.
+
+void macho_parse_fat_file(
+    void* data,
+    uint64_t size,
+    YR_OBJECT* object,
+    YR_SCAN_CONTEXT* context)
+{
+  FAT_DATA* st = (FAT_DATA*)object->data;
+
+  // All data in Mach-O fat binary header(s) is in big-endian byte order.
+
+  fat_header_t* header = (fat_header_t*)data;
+  set_integer(yr_be32toh(header->magic), object, "fat_magic");
+  set_integer(yr_be32toh(header->nfat_arch), object, "nfat_arch");
+
+  fat_arch_t* archs = (fat_arch_t*)(header + 1);
+  for (size_t i = 0; i < yr_be32toh(header->nfat_arch); ++i)
+  {
+    set_integer(yr_be32toh(archs[i].cputype),
+                object, "fat_arch[%i].cputype", i);
+    set_integer(yr_be32toh(archs[i].cpusubtype),
+                object, "fat_arch[%i].cpusubtype", i);
+    set_integer(yr_be32toh(archs[i].offset),
+                object, "fat_arch[%i].offset", i);
+    set_integer(yr_be32toh(archs[i].size),
+                object, "fat_arch[%i].size", i);
+    set_integer(yr_be32toh(archs[i].align),
+                object, "fat_arch[%i].align", i);
+
+    // Store offsets of files that are out of current block.
+    size_t offset = yr_be32toh(archs[i].offset);
+    if (size < offset)
+    {
+      st->offsets = (size_t*)yr_realloc(st->offsets,
```

Comment:
dont do realloc over the same pointer, use a temporary intermediate one and check if it's returning null. to avoid memory leaks